cmd/metal-api/internal/grpc/boot-service.go

```diff
@@ -0,0 +1,411 @@
+	// we can boot into metal-hammer again.
+	// if m != nil && m.Allocation == nil || !m.Allocation.Succeeded {
+	// 	return resp, nil
+	// }
+
+}
+
+func (b *BootService) Register(ctx context.Context, req *v1.BootServiceRegisterRequest) (*v1.BootServiceRegisterResponse, error) {
+	b.log.Infow("register", "req", req)
+	if req.Uuid == "" {
+		return nil, errors.New("uuid is empty")
+	}
+	if req.Hardware == nil {
+		return nil, errors.New("hardware is nil")
+	}
+
+	disks := []metal.BlockDevice{}
+	for i := range req.Hardware.Disks {
+		d := req.Hardware.Disks[i]
+		disks = append(disks, metal.BlockDevice{
+			Name: d.Name,
+			Size: d.Size,
+		})
+	}
+	nics := metal.Nics{}
+
+	for i := range req.Hardware.Nics {
+		nic := req.Hardware.Nics[i]
+		neighs := metal.Nics{}
+		for j := range nic.Neighbors {
+			neigh := nic.Neighbors[j]
+			neighs = append(neighs, metal.Nic{
+				Name:       neigh.Name,
+				MacAddress: metal.MacAddress(neigh.Mac),
+			})
+		}
+		nics = append(nics, metal.Nic{
+			Name:       nic.Name,
+			MacAddress: metal.MacAddress(nic.Mac),
+			Neighbors:  neighs,
+		})
+	}
+
+	machineHardware := metal.MachineHardware{
+		Memory:   req.Hardware.Memory,
+		CPUCores: int(req.Hardware.CpuCores),
+		Disks:    disks,
+		Nics:     nics,
+	}
+	size, _, err := b.ds.FromHardware(machineHardware)
+	if err != nil {
+		size = metal.UnknownSize
+		b.log.Errorw("no size found for hardware, defaulting to unknown size", "hardware", machineHardware, "error", err)
+	}
+
+	m, err := b.ds.FindMachineByID(req.Uuid)
+	if err != nil && !metal.IsNotFound(err) {
+		return nil, err
+	}
+
+	var ipmi metal.IPMI
+	if req.Ipmi != nil {
+		i := req.Ipmi
+
+		ipmi = metal.IPMI{
+			Address:    i.Address,
+			MacAddress: i.Mac,
+			User:       i.User,
+			Password:   i.Password,
+			Interface:  i.Interface,
+			BMCVersion: i.BmcVersion,
+			PowerState: i.PowerState,
+		}
+		if i.Fru != nil {
+			f := i.Fru
+			fru := metal.Fru{}
+			if f.ChassisPartNumber != nil {
+				fru.ChassisPartNumber = *f.ChassisPartNumber
+			}
+			if f.ChassisPartSerial != nil {
+				fru.ChassisPartSerial = *f.ChassisPartSerial
+			}
+			if f.BoardMfg != nil {
+				fru.BoardMfg = *f.BoardMfg
+			}
+			if f.BoardMfgSerial != nil {
+				fru.BoardMfgSerial = *f.BoardMfgSerial
+			}
+			if f.BoardPartNumber != nil {
+				fru.BoardPartNumber = *f.BoardPartNumber
+			}
+			if f.ProductManufacturer != nil {
+				fru.ProductManufacturer = *f.ProductManufacturer
+			}
+			if f.ProductPartNumber != nil {
+				fru.ProductPartNumber = *f.ProductPartNumber
+			}
+			if f.ProductSerial != nil {
+				fru.ProductSerial = *f.ProductSerial
+			}
+			ipmi.Fru = fru
+		}
+
+	}
+
+	var registerState v1.RegisterState
+	if m == nil {
+		// machine is not in the database, create it
+		name := fmt.Sprintf("%d-core/%s", machineHardware.CPUCores, humanize.Bytes(machineHardware.Memory))
+		descr := fmt.Sprintf("a machine with %d core(s) and %s of RAM", machineHardware.CPUCores, humanize.Bytes(machineHardware.Memory))
```

Comment:
Never seen this somewhere because machine skeleton is created already on PXE boot event. Can be removed.

Reply:
deleted